### PR TITLE
[ESWE-1226] Validate duplicate employers' name

### PR DESCRIPTION
**Pull request by rickchoijd**
- Validate duplicate employers' name
- Add unique constraint to employer's name

### CODE REVIEW

File: src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployerMother.kt [hunk 1]
```diff
@@ -18,7 +18,7 @@ object EmployerMother {
 
   val tescoLogistics = Employer(
     id = EntityId("2c8032bf-e583-4ae9-bcec-968a1c4881f9"),
-    name = "Tesco",
+    name = "Tesco Logistics",
```
Comment (rickchoijd) [inline src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployerMother.kt:21]: Our test data already contains duplicate employer name... thus revised.

File: src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPutShould.kt
Comment (rickchoijd) [inline src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPutShould.kt:1]: add test cases for duplicate name of employers

File: src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/employers/infrastructure/EmployerRepositoryShould.kt [hunk 2]
```diff
@@ -133,6 +133,21 @@ class EmployerRepositoryShould : EmployerRepositoryTestCase() {
...
+    }
+
+    val exception = assertFailsWith<Throwable> {
+      employerRepository.saveAndFlush(employer)
+    }
+    assertThat(exception)
+      .isInstanceOfAny(DataIntegrityViolationException::class.java)
+      .message().contains("ERROR: duplicate key value violates unique constraint")
+  }
+
```
Comment (rickchoijd) [inline src/integrationTest/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/employers/infrastructure/EmployerRepositoryShould.kt:150]: add test case for duplicate name of employers

File: src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/config/HmppsJobsBoardApiExceptionHandler.kt [hunk 3]
```diff
@@ -55,6 +62,22 @@ class HmppsJobsBoardApiExceptionHandler {
...
+        DataErrorResponse(
+          status = BAD_REQUEST,
+          timestamp = timeProvider.nowAsInstant(),
+          userMessage = "Validation failed",
+          error = "Bad request: ${e.message}",
+          details = e.errorDetails?.map { ErrorDetail.from(it) }?.toList(),
+        ),
+      )
+  }
+
```
Comment (rickchoijd) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/config/HmppsJobsBoardApiExceptionHandler.kt:80]: common handler for data validation exception (extensible for other errors, with same response structure for frontend)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPut.kt [hunk 4]
```diff
@@ -29,22 +30,53 @@ class EmployersPut(
...
+    description = "Create or update a Jobs Board Employer. Currently requires role <b>ROLE_EDUCATION_WORK_PLAN_EDIT</b>",
     responses = [
+      ApiResponse(
+        responseCode = "200",
+        description = "Employer updated",
+      ),
       ApiResponse(
         responseCode = "201",
         description = "Employer created",
       ),
```
Comment (rickchoijd) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPut.kt:43]: fix existing apidoc issues (e.g. missing `200`)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPut.kt [hunk 5]
```diff
@@ -29,22 +30,53 @@ class EmployersPut(
...
+                    }
+                  ],
+                  "timestamp": "2025-01-30T12:34:56.789Z"
+                }
+                """,
+              ),
+            ],
+          ),
+        ],
+      ),
```
Comment (rickchoijd) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPut.kt:72]: The new `400` error case of duplicate employer name (shown here as example)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPut.kt [hunk 6]
```diff
@@ -29,22 +30,53 @@ class EmployersPut(
...
-        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+        description = "Error: Unauthorised. The error status is set as the required authorisation was not provided.",
       ),
       ApiResponse(
         responseCode = "403",
-        description = "Incorrect permissions to access this endpoint",
-        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+        description = "Error: Access Denied. The error status is set as the required system role(s) was/were not found.",
       ),
     ],
```
Comment (rickchoijd) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPut.kt:81]: fix existing apidoc issues (incorrect response, should be empty body for `401` and `403`)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPut.kt [hunk 7]
```diff
@@ -56,17 +88,23 @@ class EmployersPut(
...
-      ServletUriComponentsBuilder
-        .fromCurrentRequest()
-        .path("/{id}")
-        .buildAndExpand(id)
-        .toUri(),
-    ).build()
+  ): ResponseEntity<Void> {
+    val request = createEmployerRequest.copy(id = id).also {
+      employerCreator.validate(it)
+    }
```
Comment (rickchoijd) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/controller/employers/EmployersPut.kt:94]: validate the request first (including duplicate name check)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/employers/application/EmployerCreator.kt [hunk 8]
```diff
@@ -43,6 +43,11 @@ class EmployerCreator(
 
   fun existsById(employerId: String): Boolean = employerRepository.existsById(EntityId(employerId))
 
+  fun validate(request: CreateEmployerRequest) {
+    val duplicateCount = request.let { employerRepository.countByNameIgnoreCaseAndIdNot(it.name.trim(), EntityId(it.id)) }
+    if (duplicateCount > 0) throw EmployerValidationException.duplicateEmployer(timeProvider.nowAsInstant())
+  }
```
Comment (rickchoijd) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/employers/application/EmployerCreator.kt:49]: More validation shall be added here when needed in future (and subject to refactoring for more robust/complex validation, if necessary... e.g chaining)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/employers/domain/Employer.kt [hunk 9]
```diff
@@ -16,7 +16,7 @@ data class Employer(
   @Id
   var id: EntityId,
 
-  @Column(name = "name", length = 100, nullable = false)
+  @Column(name = "name", length = 100, nullable = false, unique = true)
```
Comment (rickchoijd) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/employers/domain/Employer.kt:19]: tagging "unique" at JPA

File: src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/employers/domain/EmployerRepository.kt [hunk 10]
```diff
@@ -11,4 +11,5 @@ interface EmployerRepository : JpaRepository<Employer, EntityId> {
   fun findByNameIgnoringCaseContaining(name: String, pageable: Pageable): Page<Employer>
   fun findBySectorIgnoringCase(sector: String, pageable: Pageable): Page<Employer>
   fun findByNameContainingAndSectorAllIgnoringCase(name: String?, sector: String?, pageable: Pageable): Page<Employer>
+  fun countByNameIgnoreCaseAndIdNot(name: String, id: EntityId): Long
```
Comment (rickchoijd) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/employers/domain/EmployerRepository.kt:14]: The name matching is case-insensitive and excluding self

File: src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/shared/application/ValidationExceptions.kt
Comment (rickchoijd) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/jobsboard/api/shared/application/ValidationExceptions.kt:1]: base class for extending domain-specific errors

File: src/main/resources/db/migration/V1_7__ESWE-1226_unique_employer_name.sql
Comment (rickchoijd) [inline src/main/resources/db/migration/V1_7__ESWE-1226_unique_employer_name.sql:1]: add unique constraint to `employers`.`name`

**Review by gbmojo — APPROVED** [review]
